lib/sshutils/networking/networking.go

```diff
@@ -0,0 +1,341 @@
+	done chan struct{}
+	// killed is set to true when the process was killed forcibly.
+	killed atomic.Bool
+}
+
+// Request is a networking request.
+type Request struct {
+	// Operation is a networking operation.
+	Operation Operation
+	// Network is a network type.
+	Network string
+	// Address is a network address.
+	Address string
+	// X11Request contains additional info needed for x11 networking requests.
+	X11Request X11Request
+}
+
+// Operation is a networking operation.
+type Operation string
+
+const (
+	// NetworkingOperationDial is used to connect to a network address.
+	NetworkingOperationDial Operation = "dial"
+	// NetworkingOperationListen is used to create a local network listener.
+	NetworkingOperationListen Operation = "listen"
+	// NetworkingOperationListenAgent is used to create a local ssh-agent listener.
+	NetworkingOperationListenAgent Operation = "listen-agent"
+	// NetworkingOperationListenX11 is used to create a local x11 listener.
+	NetworkingOperationListenX11 Operation = "listen-x11"
+)
+
+// X11Config contains information used by the child process to set up X11 forwarding.
+type X11Request struct {
+	x11.ForwardRequestPayload
+	// DisplayOffset is the first display that we should try to get a unix socket for.
+	DisplayOffset int
+	// MaxDisplay is the last display that we should try to get a unix socket for, if all
+	// displays before it are taken.
+	MaxDisplay int
+}
+
+// NewProcess starts a new networking process with the given command, which should
+// be pre-configured from a ssh server context with Teleport reexec settings.
+func NewProcess(ctx context.Context, cmd *exec.Cmd) (*Process, error) {
+	// Create the socket to communicate over. We use a stream rather than datagram
+	// despite passing file descriptors because datagrams behave oddly when passed
+	// inherited by the child process. Specifically, closing the socket on the parent
+	// side does not unblock a read from the child.
+	remoteConn, localConn, err := uds.NewSocketpair(uds.SocketTypeStream)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer remoteConn.Close()
+	remoteFD, err := remoteConn.File()
+	if err != nil {
+		localConn.Close()
+		return nil, trace.Wrap(err)
+	}
+	defer remoteFD.Close()
+	cmd.ExtraFiles = append(cmd.ExtraFiles, remoteFD)
+
+	// Propagate stderr from the spawned Teleport process to log any errors.
+	cmd.Stderr = os.Stderr
+
+	proc := &Process{
+		cmd:  cmd,
+		conn: localConn,
+		done: make(chan struct{}),
+	}
+
+	if err := proc.start(ctx); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return proc, nil
+}
+
+// start the the networking process.
+func (p *Process) start(ctx context.Context) error {
+	if err := p.cmd.Start(); err != nil {
+		p.conn.Close()
+		return trace.Wrap(err)
+	}
+
+	// The child process writes errors to the parent connection for logging purposes.
+	go func() {
+		for {
+			buf := make([]byte, RequestBufferSize)
+			n, err := p.conn.Read(buf)
```

Comment:
This is a stream, you don't know where the boundaries of the message are. Why is this a stream?

Reply:
I changed it to a stream so that closing it would be reflected in the child process, rather than managing an extra fd for closure. I can revert https://github.com/gravitational/teleport/commit/dc76a7bb3f22f49baef45d59775eb0b7c4614405

Reply:
I see the issue with this now, reverting. Thanks.